### ISSUE TITLE
POST to CAN does not work

### ISSUE THREAD

**Issue by jackxujh** [gcivil-nyu-org/fall2019-cs-gy-6063-team-moonsurvivors#262]
In the test guidelines, it is mentioned that when a user posts to the CAN API, there should be a JSON response. The example POST got error in response.

## Expected Behavior
When POST to the CAN API, the user agent gets the correct response as stated in the test guidelines.

1. Send Post response `POST` data `9459720945368167357440` to `https://nyu-mercury-prod.herokuapp.com/api/can/`
    ```HTTP
    POST /api/can/ HTTP/1.1
    Content-Type: text/plain; charset=utf-8
    Host: nyu-mercury-prod.herokuapp.com
    Connection: close
    User-Agent: Paw/3.1.9 (Macintosh; OS X/10.15.1) GCDHTTPRequest
    Content-Length: 22
    
    9459720945368167357440


    ```
2. The user agent should get 
    ```JSON
    {
      "sof": 1,
      "can_id": 3,
      "rtr": 0,
      "ide": 1,
      "srr": 0,
      "extended_can_id": 0,
      "r0": 0,
      "data_length_code": 1,
      "data": 1,
      "crc_segment": 0,
      "crc_delimiter": 1,
      "ack_bit": 1,
      "ack_delimiter": 1,
      "end_of_frame": 0,
      "interframe_space": 0
     }
    ```

## Current Behavior

The response is some error:
```
HTTP/1.1 400 Bad Request
Connection: close
Server: gunicorn/19.9.0
Date: Tue, 26 Nov 2019 23:40:29 GMT
Content-Type: text/html; charset=utf-8
X-Frame-Options: SAMEORIGIN
Content-Length: 107
Via: 1.1 vegur

{"error": "Necessary data for the CAN ID type you specified was missing. The data was probably too short."}
```

Screenshot:
![Paw Screenshot](https://i.imgur.com/VqEooZN.png)

## Ideas for Improvement
Maybe the HTTP POST request is not correct, but I don't have ideas as how it is illy constructed. Maybe improve the API, or improve the guideline documents, so users can send legal requests.

## Steps to Reproduce

1. Use any REST API debug tool. For example, I am using [Paw](https://paw.cloud) on Mac.
2. Send a POST request as instructed in the test guidelines.
3. Get the response. The error will be there.
4. Change the user agent, and the problem remains.

## Context (Environment)
Tested on Mac with Paw as HTTP user agent.

**Comment by de846** [comment]
Thanks for the comments and feedback. The "data" we provided in our testing documentation worked at one point, but then when I added stronger constraints on the data consistency, the code (rightfully) said that a bad request was being sent. We forgot to go back and verify that the data we provided was still valid, when it was no longer. I have since added more exceptions to the CAN message processing, including returning the data that was decoded so the user has a chance to examine it. This whole thing is basically to support demonstration because in the future, a hardware sensor is going to be transmitting  messages constructed from a protocol scheme that should reduce the likelihood of receiving garbage data (it will essentially be just identifying the sensor, the data length, and the actual data).  Some of my fixes were included in #266. 

The CAN Decoder UI is just for demonstration as well, but I fixed a bug there too that wasn't displaying the error response on the UI (so it does now). In addition, I also use Django's `JsonResponse` class to send all responses back with errors. In addition to adding more exception types and writing much more detailed tests that show what are examples of valid and invalid CAN messages, I also include the decoded data JSON back per your suggestion.

**Comment by jackxujh** [comment]
Just for your consideration, API-wise, there is a great messaging/communication protocol called LCM which I used some while before, that is a great option for embedded devices. You can find the repository here: https://github.com/lcm-proj/lcm
